backend/pkg/modules/internal/fuzzer/tools/specs.go

```diff
@@ -1,36 +1,163 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
 package tools
 
 import (
 	"context"
 	"encoding/json"
 	"fmt"
+	"net/http"
+
+	"github.com/getkin/kin-openapi/openapi2"
+	"github.com/getkin/kin-openapi/openapi2conv"
+	"github.com/getkin/kin-openapi/openapi3"
+	"github.com/getkin/kin-openapi/routers"
+	"k8s.io/utils/strings/slices"
 
 	"github.com/openclarity/apiclarity/api/server/models"
 	"github.com/openclarity/apiclarity/backend/pkg/modules/internal/core"
+	"github.com/openclarity/apiclarity/backend/pkg/modules/internal/fuzzer/logging"
+	"gopkg.in/yaml.v2"
 )
 
-func GetAPISpecsInfo(ctx context.Context, accessor core.BackendAccessor, apiID uint) (*models.OpenAPISpecs, error) {
+// FuzzerSpecInfo An object containing info about a spec.
+type FuzzerSpecsInfo struct {
+	ProvidedSpec          string
+	ReconstructedSpec     string
+	ProvidedSpecInfo      *models.SpecInfo
+	ReconstructedSpecInfo *models.SpecInfo
+}
+
+func GetAPISpecsInfo(ctx context.Context, accessor core.BackendAccessor, apiID uint) (*FuzzerSpecsInfo, error) {
+	fuzzerSpecsInfo := FuzzerSpecsInfo{}
+
 	apiInfo, err := accessor.GetAPIInfo(ctx, apiID)
 	if err != nil {
 		return nil, fmt.Errorf("can't access to apiID (%v), error=%v", apiID, err)
 	}
 
-	specsInfo := &models.OpenAPISpecs{}
 	if apiInfo.ProvidedSpecInfo != "" {
 		specInfo := models.SpecInfo{}
 		if err := json.Unmarshal([]byte(apiInfo.ProvidedSpecInfo), &specInfo); err != nil {
 			return nil, fmt.Errorf("failed to unmarshal provided spec info. info=%+v: %v", apiInfo.ProvidedSpecInfo, err)
 		}
-		specsInfo.ProvidedSpec = &specInfo
+		fuzzerSpecsInfo.ProvidedSpecInfo = &specInfo
+	}
+
+	if apiInfo.ProvidedSpec != "" {
+		fuzzerSpecsInfo.ProvidedSpec = apiInfo.ProvidedSpec
 	}
 
 	if apiInfo.ReconstructedSpecInfo != "" {
 		specInfo := models.SpecInfo{}
 		if err := json.Unmarshal([]byte(apiInfo.ReconstructedSpecInfo), &specInfo); err != nil {
 			return nil, fmt.Errorf("failed to unmarshal reconstructed spec info. info=%+v: %v", apiInfo.ReconstructedSpecInfo, err)
 		}
-		specsInfo.ReconstructedSpec = &specInfo
+		fuzzerSpecsInfo.ReconstructedSpecInfo = &specInfo
+	}
+
+	if apiInfo.ReconstructedSpec != "" {
+		fuzzerSpecsInfo.ReconstructedSpec = apiInfo.ReconstructedSpec
+	}
+
+	return &fuzzerSpecsInfo, nil
+}
+
+func IsV2Specification(data []byte) bool {
+	specV2OrV3 := struct {
+		Openapi *string `json:"openapi"`
+		Swagger *string `json:"swagger"`
+	}{}
+	if err := yaml.Unmarshal(data, &specV2OrV3); err != nil {
```

Comment:
will it work if the input is json?

Reply:
yes. yaml.unmarshal support json AND yaml (in fact, in case of yaml, it does the yaml->json conversion)